plugins/node.d.linux/psi

```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
+
+: << '=cut'
+
+=head1 NAME
+
+psi - Plugin to graph Linux pressure stall information
+
+=head1 APPLICABLE SYSTEMS
+
+Linux kernels since 4.20 with CONFIG_PSI=y.
+
+=head1 INTERPRETATION
+
+When resources are contended, workloads experience latency spikes, throughput
+losses, and run the risk of OOM kills.
+
+Linux's PSI feature, introduced in kernel 4.20, identifies and quantifies the
+disruptions caused by resource contention.
+
+=head1 AUTHOR
+
+Copyright (C) 2022 Chris Down <chris@chrisdown.name>
+
+=head1 LICENSE
+
+GNU GPLv2
+
+=head1 MAGIC MARKERS
+
+  #%# family=auto
+  #%# capabilities=autoconf
+
+=cut
+
+psi_dir=/proc/pressure
+
+autoconf() {
+    if [ -d "$psi_dir" ]; then
+        echo yes
+    else
+        echo "no ($psi_dir not found)"
+    fi
+}
+
+config() {
+    echo "graph_title PSI"
+    echo "graph_args --upper-limit 100 -l 0"
+    echo "graph_vlabel %"
+    echo "graph_category system"
+    echo "graph_info This graph shows resource contention on the machine." \
+        "Values are recorded from the 'avg10' counter." \
+        "%full cpu is excluded as it's meaningless at the system level."
+
+    for res in memory cpu io; do
+        while read -r psi_type _; do
+            if [ "$res" = cpu ] && [ "$psi_type" = full ]; then
+                # See comment in get_data
+                continue
+            fi
+            printf '%s%spressure.label %%%s %s pressure\n' \
+                "$res" "$psi_type" "$psi_type" "$res"
+            printf '%s%spressure.info Percent %s time in "%s" contention\n' \
+                "$res" "$psi_type" "$res" "$psi_type"
+            printf '%s%spressure.type DERIVE\n' "$res" "$psi_type"
+            printf '%s%spressure.min 0\n' "$res" "$psi_type"
+        done < "$psi_dir/$res"
+    done
+}
+
+get_data() {
+    # Example line: some avg10=0.00 avg60=0.00 avg300=0.17 total=91233216
+    #
+    # CPU has "full" to keep parsers the same as for cgroup-local PSI stats,
+    # but it's meaningless at the system level, so it's ignored.
+    for res in memory cpu io; do
+        while read -r psi_type pct_ten _; do
+            if [ "$res" != cpu ] || [ "$psi_type" != full ]; then
+                pct_ten=${pct_ten#*=}
+                printf '%s%spressure.value %s\n' "$res" "$psi_type" "$pct_ten"
+            fi
+        done < "$psi_dir/$res"
+    done
+}
+
+case ${1-data} in
```

Comment:
Probably you meant `1:-` here (you missed the colon).
And please use `fetch` instead of `data` as the default (some munin implementations use this word).

Reply:
`-` seems more logical than `:-`, since if someone passes `plugin ''` that should be an error. Happy to change the verb :-)

Reply:
I appreciate your desire for distinguishing between an empty argument and an unset argument. :)

But I have a few issues with this specific implementation:
* A) it *looks* like a malformed `${foo:-default}` (thus I feel the urge to fix it)
* B) I could not find a documentation in `man dash` or `man bash` for this expression (but I tried it and it works - to my surprise). Do you have a reference?
* C) `data` looks like a magic *munin* word - but it is just a magic *plugin* word
* D) The other munin plugins are rather relaxed with regard to the cleanliness of their input. I would not try to introduce a stricter behavior solely in this plugin.

Thus I would recommend `${1:-fetch}` :)

Reply:
`-` is POSIX, see the table here: https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_06_02

If you prefer `:-` to more closely match existing plugins, that's fine. I've just sent a new version with `:-fetch`. :-)

Reply:
Hehe, thanks for sharing this piece of obscure knowledge! I always missed that before ...

Indeed: your usage of the hyphen is in this situation more correct, than my suggestion.
But I really cannot remember, having ever stumbled upon this type of parameter expansion anywhere (and I think, I have read a *lot* of shell code). Munin's `contrib` repository contains exactly two shell plugins (`hp2600_count_` and `hp2600_status_`), which use this parameter expansion.

Thus I think, we should prefer the (imperfect), but less exotic approach with `:-`. Thank you for changing this!